# HG changeset patch
# Parent 5a40c1c4ff5342afdf325819cbf5f06af6d4f6cd
# User Ben Kelly <ben@wanderview.com>
Cache.put should not mark Request/Response body used if body is not present.

diff --git a/dom/cache/TypeUtils.cpp b/dom/cache/TypeUtils.cpp
--- a/dom/cache/TypeUtils.cpp
+++ b/dom/cache/TypeUtils.cpp
@@ -189,17 +189,19 @@ TypeUtils::ToPCacheRequest(PCacheRequest
     return;
   }
 
   nsRefPtr<InternalRequest> internalRequest = aIn.GetInternalRequest();
   MOZ_ASSERT(internalRequest);
   nsCOMPtr<nsIInputStream> stream;
 
   internalRequest->GetBody(getter_AddRefs(stream));
-  aIn.SetBodyUsed();
+  if (stream) {
+    aIn.SetBodyUsed();
+  }
 
   SerializeCacheStream(stream, &aOut.body());
 }
 
 // static
 void
 TypeUtils::ToPCacheRequest(const GlobalObject& aGlobal,
                            PCacheRequest& aOut,
@@ -291,17 +293,19 @@ TypeUtils::ToPCacheResponse(PCacheRespon
 
   if (aIn.BodyUsed()) {
     aRv.ThrowTypeError(MSG_REQUEST_BODY_CONSUMED_ERROR);
     return;
   }
 
   nsCOMPtr<nsIInputStream> stream;
   aIn.GetBody(getter_AddRefs(stream));
-  aIn.SetBodyUsed();
+  if (stream) {
+    aIn.SetBodyUsed();
+  }
 
   SerializeCacheStream(stream, &aOut.body());
 }
 
 // static
 void
 TypeUtils::ToPCacheQueryParams(PCacheQueryParams& aOut, const QueryParams& aIn)
 {
